### PR TITLE
Adapt to changes in next HDF5 release

### PR DESCRIPTION
**Pull request by jmert**

### CODE REVIEW

**Comment by musm** [comment]
Should I merge this PR?  @jmert  do you have merge permission on JLD (I thought I added you).

**Comment by jmert** [comment]
I do — just thought I'd give you a chance to look through it before doing so. The last commit in particular is interesting because it's due to an apparent interaction with libhdf5 moving from v1.10 to v1.12. The reproducer I included in the commit message works on both lib versions, but for whatever reason, v1.10 doesn't trigger it in JLD's complex usage while v1.12 does.

**Comment by jmert** [comment]
CI errors look real, too. JLD tests passed, but JLDArchives doesn't. Probably some usages in the JLD00 submodule that I missed.

**Comment by musm** [comment]
Work around looks fine for now, not sure what to make of it either, but we should probably track in a new issue, so not to stall the upgrade.

**Comment by jmert** [comment]
That last commit to patch up JLD00 is maybe a bit shaky — I mostly just pattern matched and edited until the JLDArchives tests all passed. I don't know have mental model for how JLD is doing the type conversions, so I'm less confident on that commit being the right way to do things.

**Comment by musm** [comment]
That sounds fine to me. What should we do with CI on 1.0. AFAIK 1.0 is not using the latest BB archtecture on HDF5, so none of these changes are tested on that platform. I'm not sure if we should just drop testing on 1.0 or bump the min to 1.3 on Julia in Project.toml.

**Comment by jmert** [comment]
I think we bump to Julia 1.3 — that's what happened in MAT. I intended to, but just forgot that change.

**Comment by musm** [comment]
Looking at `names` on Windows. On the failing test we have:
```julia
n = ["_creator\\ENDIAN_BOM", "_creator\\JULIA_MAJOR", "_creator\\JULIA_MINOR", "_creator\\JULIA_PATCH", "_creator\\JULIA_PRERELEASE", "_creator\\WORD_SIZE", "_refs", "mygroup"]
reserved = ["_refs", "_types", "_require", "_creator"]
```
At a very quick glance:
Either `parent.plain` is adding `\\ENDIAN_BOM` unnecessarily to `_creator`  and the other values on Windows or 
https://github.com/JuliaIO/JLD.jl/blob/master/src/JLD.jl#L1129
should be instead something like  `any(occursin.(reserved, n[i]))` (altought this segfaults)

**Comment by musm** [comment]
Actually it might be due to 
```
const pathrefs = "/_refs"
const pathtypes = "/_types"
const pathrequire = "/_require"
const pathcreator = "/_creator"
```

**Comment by jmert** [comment]
I was just looking into that, and I expect it's because it's using `joinpath`: https://github.com/JuliaIO/JLD.jl/blob/e10bf154c56b71bbf7a274743ef431359043814e/src/JLD.jl#L174-L184

Since v1.3 doesn't error but v1.5 and nightly do, I expect there must have been a change to `joinpath` in Base sometime after v1.3? I was just trying to track down if that was the case.

**Comment by musm** [comment]
In any case, `joinpath` doesn't look like the right function to use here since it looks like we should be using `/` for everything

My monkey-patch solution
replace `joinpath` -> `_joinpath`

and add:
```
_joinpath(path::AbstractString)::String = path
function _joinpath(path::AbstractString, paths::AbstractString...)::String
    for p in paths
        if isabspath(p)
            path = p
        elseif isempty(path) || path[end] == '/'
            path *= p
        else
            path *= "/" * p
        end
    end
    return path
end
```

**Comment by musm** [comment]
You might also have to wrap this test https://github.com/JuliaIO/JLD.jl/blob/master/test/jldtests.jl#L965   `if !Sys.iswindows() ... end`, since mmap arrays, I think have some issues on Windows.

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -30,7 +42,12 @@ jobs:
       - run: |
           julia --project --code-coverage=user -e '
             using Pkg
-            Pkg.GitTools.clone(Pkg.Types.Context(), "https://github.com/JuliaIO/JLDArchives.jl.git", "JLDArchives")
+            url = "https://github.com/JuliaIO/JLDArchives.jl.git"
```
Comment (musm) [inline .github/workflows/CI.yml:45]: I think all of the quotes here need escaping.

Comment (jmert) [reply .github/workflows/CI.yml:45]: I was just going to ask if it was a quoting issue on Windows.

**Comment by musm** [comment]
Finally!

**Comment by jmert** [comment]
I'll squash a few of the commits again, since I'd like to keep type comparison workaround commit separate for easier historical debugging.

**Pull request merged**